mesonbuild/compilers.py

```diff
@@ -447,22 +447,34 @@ def _should_prepend(cls, arg):
             return True
         return False
 
-    def to_native(self):
+    def _fix_broken_linkers(self):
         # Check if we need to add --start/end-group for circular dependencies
         # between static libraries.
-        if get_compiler_uses_gnuld(self.compiler):
-            group_started = False
-            for each in self:
-                if not each.startswith('-l') and not each.endswith('.a'):
-                    continue
-                i = self.index(each)
-                if not group_started:
-                    # First occurance of a library
-                    self.insert(i, '-Wl,--start-group')
-                    group_started = True
-            # Last occurance of a library
-            if group_started:
-                self.insert(i + 1, '-Wl,--end-group')
+        if not get_compiler_uses_gnuld(self.compiler):
+            return
+        group_started = False
+        # Must ensure that we don't accidentally nest --start/end-group
+        # since older `ld` versions (such as 2.20) barf on that:
+        # https://github.com/mesonbuild/meson/issues/1936
+        # We could add complicated logic here to detect nesting, but
+        # users should not be adding --start-group themselves. KISS for now.
+        for each in self:
+            if each.endswith('--start-group'):
+                return
+        for each in self:
+            if not each.startswith('-l') and not each.endswith('.a'):
+                continue
+            i = self.index(each)
+            if not group_started:
+                # First occurance of a library
+                self.insert(i, '-Wl,--start-group')
+                group_started = True
+        # Last occurance of a library
+        if group_started:
+            self.insert(i + 1, '-Wl,--end-group')
```

Comment:
Actually, this is not broken because of undefined behaviour. We are iterating over an array we are modifying, and we end up iterating over the `-lfoo` twice. This might break in the future, so I'll change this so we iterate over a copy of the array.